core/block_validator.go

```diff
@@ -137,9 +140,73 @@ func (v *BlockValidator) ValidateBody(block *types.Block) error {
 	return nil
 }
 
+// ValidateSystemTxs validates all system txs contained in a block.
+// We check that:
+// - the sender is a predetermined address
+// - the recipient is a system contract
+func (v *BlockValidator) ValidateSystemTxs(block *types.Block) error {
+	// first pass: ensure that system txs are first, in a continuous block
+	haveSystemTx := false
+	haveNonSystemTx := false
+	for _, tx := range block.Transactions() {
+		if tx.IsSystemTx() {
+			if !v.config.Scroll.SystemTxEnabled() {
+				return ErrSystemTxNotEnabled
+			}
+
+			if haveNonSystemTx {
+				return consensus.ErrInvalidL1MessageOrder
+			}
+
+			haveSystemTx = true
+			continue
+		}
+
+		haveNonSystemTx = true
+	}
+
+	if !haveSystemTx {
+		return nil
+	}
+
+	for _, tx := range block.Transactions() {
+		if !tx.IsSystemTx() {
+			break
+		}
+
+		stx := tx.AsSystemTx()
+
+		found := false
+		for _, signer := range v.config.Scroll.SystemTx.Senders {
+			if stx.From == signer {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return ErrUnknownSystemSigner
+		}
+
+		found = false
+		for _, contract := range v.config.Scroll.SystemTx.Contracts {
```

Comment:
You said you wanted to move this contract list into state? Is that left for a future PR?

Another way is to have a static list (not in config), that can be extended at hard forks. You can check how the precompiles and instruction set are configured.

Reply:
Aha, static list seems best.

Reply:
added bonus that we have clear system service activation rules at fork boundaries.

Reply:
will change to use the static list and drop the error prone dependence on config.